### ISSUE TITLE
Question ?

### ISSUE THREAD

**Issue by melck** [mridgers/clink#329]
hi,

I'm trying to create a match generator who autocomplete folder like ide done (ex atom sublime). I developed a command program made with nodejs to search directories.
I want to rewrite output line but i don't know how to do it with lua (maybe impossible). Do u have a solution even a ugly way to rewrite ouput with lua ?

ex: "ws pro test" would become "ws project-test"

```Lua
function ws_match_generator(text, first, last)
	found_matches = false
	if rl_state.line_buffer:find("^ws ") then
		clink.match_display_filter = ws_display_filter
		found_matches = true
		
		local matches = {}
		local i = 1
		local fragment = rl_state.line_buffer:match('^ws%s+(.*)')
		for line in io.popen("find-project search \"" .. fragment .. "\""):lines() do
			matches[i] = line
			i = i + 1
		end

		if (#matches == 1) then
			-- it's here that i want to rewrite output line
			rl_state.line_buffer:gsub('^(.*)', matches[1]);
		elseif (#matches > 1) then
			for i= 0, #matches do
				clink.add_match(matches[i])
			end
		end
	end

    return found_matches
end
```

PS: sorry for my poor english :)

**Comment by mridgers** [comment]
First up I noticed that the first thing your Lua script does is look to see if you're generating matches for a specific command ("ws" in this case). Clink has a framework for this where you build a parser that describes a particular command's arguments and flags and register it with Clink;

```lua
local ws_parser = clink.arg.new_parser()
ws_parser:set_flags( ... )
ws_parser:set_arguments( ... )
clink.arg.register_parser("ws", ws_parser)
```

There's more details in the [Clink's documentation](https://github.com/mridgers/clink/blob/master/docs/clink.md#argument-completion)

As for you question, I'm afraid what you're trying to do isn't really possible because the match generation is only for the word under the cursor but "pro test" is two words. There's another complication which is that if you're using "clink.arg" (which you're not currently) then it is quite picky about what it accepts as a match.

Having said that, you can get sort of close, which may or may not be acceptable depending what you're trying to do;

```lua
clink.is_match = function () return true end

function ws_first_word_func(word)
	return { "project-test" }
end

local ws_parser = clink.arg.new_parser()
ws_parser:set_arguments({ ws_first_word_func })

clink.arg.register_parser("ws", ws_parser)
```

This will replace the first word/argument of an entered "ws" command with "project-test";

```
c:\>ws abc<TAB>
c:\>ws project-test _
```

**Comment by melck** [comment]
Thank you for your advices. My purpose is to fuzzy filtering my projects folders by using a command tool which i have developed with nodejs and fuzzy-filtering's tool used by Atom

If i understand correctly, maybe be the next syntax could answer to my issue ?
```
c:\>ws "project abc<tab>
```

I have another problem to use my command, how can i force order of multiple matches ?

**Comment by mridgers** [comment]
>maybe be the next syntax could answer to my issue ?
Correct. The quote character will mean that the 'project abc' will be considered as a single word to be completed.

>I have another problem to use my command, how can i force order of multiple matches ?
I'm sure I fully understand what you mean? It isn't possible to reorder words on a line from Lua for the reasons I mentioned before. Clink's Lua-base framework is just for generating matches for the word under the cursor - it isn't possible to rewrite the line being edited.
